xADCSCertificationAuthority - Fix Get-TargetResource + HQRM cleanup and tests - Fixes #14

PlagueHO opened the pull request:
This PR addresses the issue raised in #14 .

It also includes the following changes:
- xAdcsCertificateAuthority:
  - Change property format in Readme.md to be standard layout.
  - Converted style to meet HQRM guidelines.
  - Added verbose logging support.
  - Added string localization.
  - Fixed Get-TargetResource by removing IsCA and changing Ensure to return whether or not CA is installed.
  - Added unit tests.
- xAdcsOnlineResponder:
  - Change property format in Readme.md to be standard layout.
  - Added unit test header to be latest version.
- xAdcsWebEnrollment:
  - Change property format in Readme.md to be standard layout.
  - Added unit test header to be latest version.
- Added CommonResourceHelper.psm1 (copied from xPSDesiredStateConfiguration).
- Removed Technet Documentation HTML file from root folder.
- Removed redundant code from AppVeyor.yml.

I have added integration tests, but they can't be run on AppVeyor yet because installing ADCS on WS2012R2 requires a reboot - WS2016 does not seem to so they should be able to be run on that when it is available as an AppVeyor node.

You can run the integration tests manually though on a WS2012R2 machine with ADCS installed. You do need to provide credentials to the integraiton test script however (set in variables). This will be able to automated in WS2016.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xadcsdeployment/19)

<!-- Reviewable:end -->

Indhukrishna commented on DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1:
```diff
@@ -1,162 +1,485 @@
+    catch
+    {
+        # CA is probably already installed - output the exception just in case
+        # it is some other problem.
```
If it is  possible to know the exact exception that we will get when CA is already installed, is it not better to look for that explicitly and use that as the criteria for this case?

PlagueHO replied:
Good call. I've managed to figure out the exception that will be thrown and am catching that now.

Indhukrishna commented on Tests/Unit/MSFT_xAdcsWebEnrollment.Tests.ps1:
```diff
@@ -65,7 +62,7 @@ try
         }
 
-        Describe "$($Global:DSCResourceName)\Set-TargetResource" {
+        Describe 'MSFT_xAdcsWebEnrollment\Set-TargetResource' {
```
>'MSFT_xAdcsWebEnrollment [](start = 17, length = 24)

Can we not use the variable here?

PlagueHO replied:
I've converted these back to variables (but not $global or $script).

Indhukrishna commented on Tests/Integration/MSFT_xAdcsCertificationAuthority.Tests.ps1:
```diff
@@ -0,0 +1,142 @@
+    # Ensure that the tests can be performed on this computer
+    if (-not (Test-WindowsFeature -Name 'ADCS-Cert-Authority'))
+    {
+        Return
```
>Return [](start = 8, length = 6)

A Verbose message here would be helpful for diagnosability.

PlagueHO replied:
The ```Test-WindowsFeature``` function writes a verbose message already saying:
```Integration tests cannot be run because ADCS-Cert-Authority is not installed.```

PlagueHO commented:
Thank you so much @Indhukrishna for reviewing. :+1: I'll get the changes made when I get home from work tonight and have them ready for Friday morning PDT.

PlagueHO commented:
Review status: 0 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1, line 142 at r1](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KTXgpt0uYXnScEKEXFJ-r1-142:-KU1-uMOiVojI_MPsc_7:b-896fix) ([raw file](https://github.com/powershell/xadcsdeployment/blob/5ef6d406db5cb95e01af03845621445c75a241d2/DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1#L142)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary>
> Good call. I've managed to figure out the exception that will be thrown and am catching that now.
</details>
Done.

---

*[Tests/Integration/MSFT_xAdcsCertificationAuthority.Tests.ps1, line 11 at r1](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KTXgpt0uYXnScEKEXFT-r1-11:-KU1-wfWlsegwda1RhNm:b-896fix) ([raw file](https://github.com/powershell/xadcsdeployment/blob/5ef6d406db5cb95e01af03845621445c75a241d2/Tests/Integration/MSFT_xAdcsCertificationAuthority.Tests.ps1#L11)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary>
> Oh wow! Ok, I didn't know that. I'll see if I can get the integration tests working in AppVeyor.
</details>
Done.

---

*[Tests/Unit/MSFT_xAdcsWebEnrollment.Tests.ps1, line 65 at r1](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KTXgpt0uYXnScEKEXFN-r1-65:-KU103NVPRoTBS1KcWGr:b-896fix) ([raw file](https://github.com/powershell/xadcsdeployment/blob/5ef6d406db5cb95e01af03845621445c75a241d2/Tests/Unit/MSFT_xAdcsWebEnrollment.Tests.ps1#L65)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary>
> I've converted these back to variables (but not $global or $script).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
@Indhukrishna - I think I've completed everything.

Thanks again for the tip on AppVeyor reboots - that is definitely a fantastic tip - we should be able to now add Integration tests to the xActiveDirectory resource! 😁

PlagueHO commented:
Hi @mbreakey3 - I've actually finished the changes requested for this PR. So just waiting on a review.

PlagueHO commented:
@kwirkykat, @mbreakey3, @Indhukrishna - I don't suppose you have anyone over there who can complete this review?

It would be really good to get this one sorted for @theJasonHelmick and @majst32 .

PlagueHO commented:
PS: I've updated the resource parameter formats to meet the new parameter format guidelines determined a couple of days ago. I've also removed all the markdown violations from the readme.md.

PlagueHO commented:
This PR now also:
- Fixes #20 
- Fixes #21 

I may also try and get Integration tests set up for these remaining resources in this PR. This should then make this module HQRM.

mgreenegit commented:
Reviewed 7 of 16 files at r1, 2 of 6 files at r2, 1 of 4 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[appveyor.yml, line 8 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWA-9Oe8YspMLNtLdXi:-KWA-9Oe8YspMLNtLdXj:bm6fvq) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/appveyor.yml#L8)):*
> ```YAML
>   - git clone https://github.com/PowerShell/DscResource.Tests
>   - ps: |
>         Import-Module .\DscResource.Tests\TestHelper.psm1 -force
> ```

Add parameter name

---

*[appveyor.yml, line 12 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWA-iEDXSppDXWYoPcc:-KWA-iEDXSppDXWYoPcd:bn87kr8) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/appveyor.yml#L12)):*
> ```YAML
>         Install-Module -Name Pester -Repository PSGallery -Force
>         Install-WindowsFeature -Name ADCS-Cert-Authority,RSAT-ADCS
>         & net @('user','AdcsAdmin','NotPass12!','/ADD')
> ```

Could this be replaced by new-localuser in 5.1?
https://msdn.microsoft.com/en-us/powershell/reference/5.1/microsoft.powershell.localaccounts/new-localuser

---

*[appveyor.yml, line 13 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWA-ral9SO_noXg95hc:-KWA-ral9SO_noXg95hd:b-4kl1y7) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/appveyor.yml#L13)):*
> ```YAML
>         Install-WindowsFeature -Name ADCS-Cert-Authority,RSAT-ADCS
>         & net @('user','AdcsAdmin','NotPass12!','/ADD')
>         & net @('localgroup','administrators','AdcsAdmin','/ADD')
> ```

Could this be replaced with Add-LocalGroupMember in 5.1?
https://msdn.microsoft.com/en-us/powershell/reference/5.1/microsoft.powershell.localaccounts/add-localgroupmember

---

*[DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1, line 192 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWAK6a-mCPIDZteCC57:-KWAK6a-mCPIDZteCC58:b2lvk3q) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1#L192)):*
> ```PowerShell
>     }
> 
>     return @{
> ```

Align on "="

---

*[Tests/Unit/MSFT_xAdcsOnlineResponder.Tests.ps1, line 13 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWAKVaw7YG5Nl3pG0OV:-KWAKVaw7YG5Nl3pG0OW:bm6fvq) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/Tests/Unit/MSFT_xAdcsOnlineResponder.Tests.ps1#L13)):*
> ```PowerShell
> }
> 
> Import-Module (Join-Path -Path $script:moduleRoot -ChildPath 'DSCResource.Tests\TestHelper.psm1') -Force
> ```

Add parameter name

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Review status: 12 of 18 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[appveyor.yml, line 8 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWA-9Oe8YspMLNtLdXi:-KWC0u-lahgAU-1MSNYJ:b-896fix) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/appveyor.yml#L8)):*
<details><summary><i>Previously, mgreenegit (Michael Greene) wrote…</i></summary>
> Add parameter name
</details>
Done.

---

*[appveyor.yml, line 12 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWA-iEDXSppDXWYoPcc:-KWC0uecYsURgzqk52Pg:bfp8asl) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/appveyor.yml#L12)):*
<details><summary><i>Previously, mgreenegit (Michael Greene) wrote…</i></summary>
> Could this be replaced by new-localuser in 5.1?
> https://msdn.microsoft.com/en-us/powershell/reference/5.1/microsoft.powershell.localaccounts/new-localuser
</details>
Really good idea - but unfortunately (as I found out) the AppVeyor build workers are still at WMF 5.0. So will need to wait for a WMF5.1 build worker to be available. I raised an issue though so that when it is available I can convert it.

---

*[appveyor.yml, line 13 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWA-ral9SO_noXg95hc:-KWC1BweHLnPctDtbWZi:b-q64pkt) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/appveyor.yml#L13)):*
<details><summary><i>Previously, mgreenegit (Michael Greene) wrote…</i></summary>
> Could this be replaced with Add-LocalGroupMember in 5.1?
> https://msdn.microsoft.com/en-us/powershell/reference/5.1/microsoft.powershell.localaccounts/add-localgroupmember
</details>
See above.

---

*[DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1, line 192 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWAK6a-mCPIDZteCC57:-KWC-H4jWGXejRqnWExE:boj1qgt) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1#L192)):*
<details><summary><i>Previously, mgreenegit (Michael Greene) wrote…</i></summary>
> Align on "="
</details>
Good catch. Done.

---

*[Tests/Unit/MSFT_xAdcsOnlineResponder.Tests.ps1, line 13 at r4](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19#-KWAKVaw7YG5Nl3pG0OV:-KWC-bQIsMkxKzABfddA:bsojjcf) ([raw file](https://github.com/powershell/xadcsdeployment/blob/1d4835b7e87dd956a5ff9331941c973043857e04/Tests/Unit/MSFT_xAdcsOnlineResponder.Tests.ps1#L13)):*
<details><summary><i>Previously, mgreenegit (Michael Greene) wrote…</i></summary>
> Add parameter name
</details>
Great catch. I also fixed the ones in the other unit tests and the integration test. This actually comes from the unit/integration test template. So I'll get a PR together to fix this in the template as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xadcsdeployment/19)*
<!-- Sent from Reviewable.io -->

PlagueHO commented on DSCResources/CommonResourceHelper.psm1:
```diff
@@ -0,0 +1,142 @@
+    [CmdletBinding()]
+    param ()
+
+    return $PSVersionTable.PSEdition -ieq 'Core'
```
Thanks @TravisEz13 - I can't actually track down the code you're referring to - I thought it was in xTimezone, but I can't find it anywhere. Do you recall where it was?

However, I actually copied this code straight out of the new [PSDscResources](https://github.com/PowerShell/PSDscResources/blob/dev/DscResources/CommonResourceHelper.psm1#L5) module, so if this code is wrong, we should change it in PSDscResources as well (it is used in [xPSDesiredStateConfiguration](https://github.com/PowerShell/xPSDesiredStateConfiguration/blob/dev/DSCResources/CommonResourceHelper.psm1#L5) too).

PlagueHO replied:
@TravisEz13 - sorry - I found where you'd put the correct code (I hadn't got to your other code review yet). I'll raise an issue against the PSDscResources and xPSDesiredStateConfiguration repos to get these updated to use the correct method.

PlagueHO replied:
I have updated this now. Cheers.

PlagueHO commented:
@mgreenegit - would you be able to go into reviewable and approve the changes you requested for this PR? Then I can merge this PR. Thank you - I appreciate you taking the time for the review of this one - it was very helpful.

PlagueHO commented:
Note: I merged this even though @mgreenegit hadn't signed off the requested changes. I did make the changes he requested though.
